src/test/java/bio/terra/integration/DataRepoClient.java

```diff
@@ -87,37 +87,66 @@ public DataRepoClient() {
+
+        // if the initial response is bad gateway, then the request probably never got delivered
+        if (jobModelResponse.getStatusCode() == HttpStatus.BAD_GATEWAY) {
+            throw new IllegalStateException("unexpected job status code: " + jobModelResponse.getStatusCode());
+        }
+
+        boolean keepGoing = true;
+        String location = getLocationHeader(jobModelResponse);
+        while (keepGoing) {
+            switch (jobModelResponse.getStatusCode()) {
+                case ACCEPTED: {
+                    jobModelResponse = waitForResponseUntilNot(user, location, HttpStatus.ACCEPTED);
+                    break;
+                }
+
+                case BAD_GATEWAY:
+                    jobModelResponse = waitForResponseUntilNot(user, location, HttpStatus.BAD_GATEWAY);
+                    break;
+
+                default:
+                    keepGoing = false;
+                    break;
+            }
+        }
+
+        if (jobModelResponse.getStatusCode() != HttpStatus.OK) {
+            throw new IllegalStateException("unexpected job status code: " + jobModelResponse.getStatusCode());
+        }
+
+        location = getLocationHeader(jobModelResponse);
+        return get(user, location, responseClass);
+    }
+
+    // poll for a response with a return status that is not specified status
+    private DataRepoResponse<JobModel> waitForResponseUntilNot(TestConfiguration.User user,
+                                                               String location,
+                                                               HttpStatus notStatus) throws Exception {
         final int initialSeconds = 1;
         final int maxSeconds = 16;
 
         try {
             int count = 0;
             int sleepSeconds = initialSeconds;
-            while (jobModelResponse.getStatusCode() == HttpStatus.ACCEPTED) {
-                String location = getLocationHeader(jobModelResponse);
-                logger.info("try #{} for {}", ++count, location);
 
+            while (true) {
+                logger.info("try #{} until not {} for {}", ++count, notStatus, location);
                 TimeUnit.SECONDS.sleep(sleepSeconds);
-                jobModelResponse = get(user, location, JobModel.class);
-
-                int nextSeconds = 2 * sleepSeconds;
-                sleepSeconds = (nextSeconds > maxSeconds) ? maxSeconds : nextSeconds;
+                sleepSeconds = Math.min(2 * sleepSeconds, maxSeconds);
+
+                DataRepoResponse<JobModel> jobModelResponse = get(user, location, JobModel.class);
+                logger.info("Got response. status: " + jobModelResponse.getStatusCode()
+                    + " location: " + jobModelResponse.getLocationHeader().orElse("not present"));
+                if (jobModelResponse.getStatusCode() != notStatus) {
+                    return jobModelResponse;
+                }
             }
         } catch (InterruptedException ex) {
```

Comment:
I encountered some IOExceptions here when I was doing my Kubernetes testing. Do you think that would be worth handling here?

Reply:
Sure. What do you do when you catch one? Can I copy some code??

Reply:
Thinking about it, this only makes sense if you want to add some sort of retry mechanism. I followed the pattern in GoogleResourceService -> enableIamPermissions.